extensions/ql-vscode/src/databases-ui.ts

```diff
@@ -360,6 +369,45 @@ export class DatabaseUI extends DisposableObject {
     }
   };
 
+  handleRemoveOrphanedDatabases = async (): Promise<void> => {
+    logger.log('Removing orphaned databases.');
+    let dbDirs =
+      // read directory
+      (await fs.readdir(this.storagePath, { withFileTypes: true }))
+        // remove non-directories
+        .filter(dirent => dirent.isDirectory())
+        // get the full path
+        .map(dirent => path.join(this.storagePath, dirent.name))
+        // filter databases still in workspace
+        .filter(dbDir => {
+          const dbUri = Uri.file(dbDir);
+          return this.databaseManager.databaseItems.every(item => item.databaseUri.fsPath !== dbUri.fsPath);
+        });
+
+    // filter non-databases
+    dbDirs = await asyncFilter(dbDirs, isDatabaseDirectory);
+
+    if (!dbDirs.length) {
+      logger.log('No orphaned databases found.');
+      return;
+    }
+
+    // delete
+    const failures = [] as string[];
+    await Promise.all(
```

Comment:
I think [`Promise.allSettled`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) gives you similar result collection behaviour without having to insert your own `catch`. You would have to check for `status == 'rejected'` to collect the failures.

Reply:
`allSettled` was introduced in node 12.9. I _think_ that's fine to use in vscode (my instance uses v12.14.1).

Reply:
Having trouble getting tsc to recognize `allSettled` in the editor. I think I need to change the tsconfig somehow, but not exactly sure.

Though, I don't think the results would be much cleaner since I still need to keep filter the results and keep track of paths.

Reply:
No worries. What you've got is pretty clear :)